Discuss manubot 0.2 features & default CSL style

dhimmel opened the pull request:
Discusses Manubot features from the 0.2 release as per https://github.com/greenelab/meta-review/issues/97

dhimmel committed:
Discuss manubot cite CLI, CSL schema validation, manubot CSL style

Discusses Manubot features from the 0.2 release as per
https://github.com/greenelab/meta-review/issues/97

agitter commented on content/02.main-text.md:
```diff
@@ -143,6 +144,25 @@ Styles define how references are constructed from bibliographic metadata, contro
 Thousands of journals have [predefined styles](http://editor.citationstyles.org/searchByName/).
 As a result, adopting the specific bibliographic format required by a journal usually just requires specifying the style's source URL in the Manubot configuration.
 
+To make citation-by-identifier easily usable outside of Manubot manuscripts, we created the `manubot cite` command line utility, available as a [Python package](https://pypi.org/project/manubot/).
```
I like highlighting the new features in this paragraph, but the syntax examples break up the flow.  What are the benefits of including them here instead of as part of the Manubot package documentation?

slochower replied:
Agree.

dhimmel replied:
Is https://github.com/greenelab/meta-review/commit/c8ce9e526e8623d384ae0d6698f7af72384bba31 better?

I think `manubot cite` utility is important to have in manuscript... it could end up being the most important contribution of our project in the longterm. It puts citation-by-identifier and easy CSL Data generation at everyone's fingertips. Previously, there was no easy way to do this, especially for all the citation sources Manubot supports.

agitter commented on content/02.main-text.md:
````diff
@@ -143,6 +144,25 @@ Styles define how references are constructed from bibliographic metadata, contro
+  pmid:29618526 doi:10.1038/550143a
+```
+
+If the user does not specify a CSL style, the program defaults to using the Manubot style, which focuses on information relevant for modern scholarly workflows and omits the rest.
````
This paragraph reads more like documentation rather than information that will get readers excited about Manubot and help them understand what it contributes to open collaborative writing.

slochower replied:
Agree, especially about the CSL. Having a default CSL is probably not part of the core message.

vsmalladi replied:
Agree, I think that we can remove this. The highlight is stated in the sentence before the example, that we want to highlight how easy it is to change the citation style.

dhimmel replied:
I agree it was too doc-like. Shortened considerably in https://github.com/greenelab/meta-review/pull/99/commits/c8ce9e526e8623d384ae0d6698f7af72384bba31. My motivation for discussing our CSL style is that I think it is innovative, and this is the right manuscript to lay claim to these innovations, which hopefully will become more widespread in the future. The point is not having _a_ default, but the point is that we _also_ created a style that innovates beyond traditional reference styles.

dhimmel replied:
P.S. I also wanted to cite my blog post on numeric versus author citation styles. This was a bit of original research that was largely motivated by Manubot... I wanted to get to the bottom of whether we should use numeric or author-style, and what the differences were.

agitter commented on content/02.main-text.md:
```diff
@@ -121,6 +121,7 @@ All citations point to a standard identifier, for which Manubot automatically re
 Table @tbl:citations presents the supported identifiers and example citations before and after Manubot processing.
 Authors can optionally define citation tags to provide short readable alternatives to the citation identifiers.
 Metadata is exported to [Citation Style Language (CSL) JSON Items](http://citeproc-js.readthedocs.io/en/latest/csl-json/markup.html#items), an open standard that is widely supported by reference managers [@doi:10.1007/978-3-319-00026-8_8; @doi:10.1080/02763869.2012.641841].
+Manubot ensures its CSL Items validate according to the official JSON schema by removing problematic fields, which are prone to crashing downstream citation processors, such as [pandoc-citeproc](http://hackage.haskell.org/package/pandoc-citeproc) in Manubot's case.
```
I accidentally dropped a comment from my review

Can we restate this to emphasize the most important parts for a general audience?
> Manubot ensures its CSL Items validate and removes problematic fields, which are prone to crashing downstream citation processors.

dhimmel replied:
Rephrased in https://github.com/greenelab/meta-review/pull/99/commits/cf1cacc3041b1f006a2fa75ceacb140276c54afd

dhimmel committed:
Add CSL Naturejobs Techblog citation

dhimmel commented on content/citation-tags.tsv:
```diff
@@ -1,4 +1,5 @@
 tag	citation
+techblog-csl	url:http://blogs.nature.com/naturejobs/2017/05/03/techblog-create-the-perfect-bibliography-with-the-csl-editor/
```
In a follow-up commit, I want to switch other underscores in tag names to `-`, as the later is valid in strict pandoc formatting, and therefore should be preferred.

dhimmel replied:
Done in https://github.com/greenelab/meta-review/pull/102

dhimmel committed:
Rephrase CSL pruning sentence & CSL Data Items

Previousely, we used the term "CSL Items". However, many places
use "CSL Data". Perhaps, it's most clear if we do "CSL Data Items".
Then we can use CSL Data as a shorthand. Not sure what the best
capitalization is.

dhimmel committed:
Move & shorten `manubot cite` paragraph

dhimmel commented:
@agitter @vsmalladi @slochower ready for re-review

agitter commented on content/02.main-text.md:
```diff
@@ -120,7 +120,9 @@ Manubot includes an additional layer of citation processing, currently unique to
 Authors can optionally define citation tags to provide short readable alternatives to the citation identifiers.
-Metadata is exported to [Citation Style Language (CSL) JSON Items](http://citeproc-js.readthedocs.io/en/latest/csl-json/markup.html#items), an open standard that is widely supported by reference managers [@doi:10.1007/978-3-319-00026-8_8; @doi:10.1080/02763869.2012.641841].
+Metadata is exported to [Citation Style Language (CSL) JSON Data Items](http://citeproc-js.readthedocs.io/en/latest/csl-json/markup.html#items), an open standard that is widely supported by reference managers [@doi:10.1007/978-3-319-00026-8_8; @doi:10.1080/02763869.2012.641841].
+Sometimes however, external resources provide Manubot with invalid CSL Data, which can crash downstream citation processors, such as [pandoc-citeproc](http://hackage.haskell.org/package/pandoc-citeproc).
```
How about changing to `However, sometimes external...`?

agitter commented on content/02.main-text.md:
```diff
@@ -140,7 +142,7 @@ In this example, the tag `Avasthi2018_preprints` represents the DOI identifier `
 As a result, users can choose from thousands of existing CSL styles or use Manubot's default style.
 Styles define how references are constructed from bibliographic metadata, controlling layout details such as the maximum number of authors to list per reference.
-Thousands of journals have [predefined styles](http://editor.citationstyles.org/searchByName/).
+Thousands of journals have [predefined styles](http://editor.citationstyles.org/searchByName/) [@tag:techblog-csl].
```
Good addition

agitter commented on content/02.main-text.md:
```diff
@@ -193,6 +195,21 @@ Additional research studies in progress are being authored using Manubot, spanni
 
+### Citation utility
+
+To make citation-by-identifier easily usable outside of Manubot manuscripts, we created the `manubot cite` command line utility, available as a [Python package](https://pypi.org/project/manubot/).
```
Do we need `available as a [Python package](https://pypi.org/project/manubot/)` here?  We already introduce the Python package at the start of the `Manubot` section.  We may want to instead change that link to PyPI, or at least make them consistent with GitHub or PyPI.

dhimmel replied:
I like the PyPI link here, because if someone wants to install the utility, that is the most useful link.

I am now not content with the previous statement:

> Manubot is a system for writing scholarly manuscripts via GitHub that is built upon our Python package of the same name.

Alone this sentences is fine, but we don't mention the Rootstock template anywhere. I am thinking we ought to mention somewhere that users currently setup a manuscript based on our template repository.

slochower replied:
Do you think we should add a paragraph introducing Rootstock?

agitter replied:
Currently we have
> An [example repository](https://github.com/greenelab/manubot-rootstock) demonstrates Manubot’s features and serves as a template for users to write their own manuscript with Manubot.

We could say more, but would it end up being the technical steps to using the rootstock?

slochower replied:
If we wanted to add a bit more about Rootstock, I would prefer it be non-technical, e.g., Rootstock is a way for non-technical folks to get started using Manubot. Clone the repository, follow the instructions, edit some Markdown, and CI will automatically render the manuscript for you. It's a complement and contrast to the power and utility of the command line `manubot` utilities.

On consideration, maybe I mean "less technical" not "non-technical" because, of course, you still do need to know `git` and you have to do some Travis stuff. But consider the case of collaborators. If someone has already setup a Rootstock repository, then all a new user has to do is edit some Markdown, possibly even in the GitHub web interface, and Manubot will build the manuscript automatically. So I see Rootstock as a way to make it easier for people to take advantage of using Manubot, even if they don't have the facility to work with the command line.

agitter replied:
Yes, I would support adding more along those lines.  I'd say that some technical expertise is required to create a new manuscript using the rootstock based on the user reports we've seen.  But the template and step-by-step instructions make it _much_ easier than trying to set up something comparable independently.

As you note, once it is set up there is much less technical expertise required for collaborators.  It does help if the manuscript maintainer understands git well enough to help resolve workflow issues and conflicts.

Should we move this discussion to a new issue?

dhimmel replied:
See issue https://github.com/greenelab/meta-review/issues/104 opened by @agitter. Decoupling this issue from the scope of this PR.

slochower replied:
Agree.

agitter commented on content/02.main-text.md:
````diff
@@ -193,6 +195,21 @@ Additional research studies in progress are being authored using Manubot, spanni
+This utility takes a list of citations and returns either a rendered bibliography or CSL Data Items (i.e. JSON-formatted reference metadata).
+For example, the following command outputs a Markdown reference list for the two specified articles according to the bibliographic style of _PeerJ_:
+
+```sh
````
This inline example still seems a little out of place to me, but if you think it is important I can live with it.  I do agree that the `manubot cite` utility is an important contribution that we don't want to have overlooked.  This emphasizes it.

dhimmel replied:
I like the code example and think we should keep it for now... would be okay removing it in future, especially if we include a graphic showing some of the citation features (or if a journal allows video and we include a terminal recording or something).

agitter commented on content/02.main-text.md:
````diff
@@ -193,6 +195,21 @@ Additional research studies in progress are being authored using Manubot, spanni
+  pmid:29618526 doi:10.1038/550143a
+```
+
+If the user does not specify a CSL style, the utility defaults to using the Manubot style, which focuses on information relevant for modern scholarly workflows and omits the rest.
````
The phrase `information relevant for modern scholarly workflows and omits the rest` seems quite subjective and isn't well-supported in this paragraph.  If there was widespread agreement about what information is relevant, we would have more homogeneous citation styles.  If we are arguing that the choices in Manubot's CSL are innovative and better than other options, we should explain that more explicitly, as your linked blog does for numeric citations.

This information could be briefly integrated into other parts of the manuscript.  In a section above we have
> As a result, users can choose from thousands of existing CSL styles or use Manubot's default style.

If the numeric citations are important, we could change that to
> As a result, users can choose from thousands of existing CSL styles or use Manubot's default style, which emphasizes titles instead of authors and numeric-style citations. [@url:https://blog.dhimmel.com/citation-styles/].

slochower replied:
I agree with @agitter. I've read the blog post and it makes the case that both styles have advantages. I prefer @agitter's final version.

dhimmel replied:
See https://github.com/greenelab/meta-review/pull/99/commits/ee1a34760e6433afc6a9e8eef95fbe07a6aa987f. I realized there was some redundancy that had crept into the CSL style paragraph, so I addressed that as well.

dhimmel committed:
Merge branch 'master' into cite-cli

dhimmel committed:
However sometimes

dhimmel committed:
Rephrase and condense CSL Style paragraph

dhimmel commented:
From talking with @adam3smith at FORCE2018, who's written thousands (I think) of CSL Styles, the most common way to refer to CSL citation metadata is "CSL JSON". Currently, we use CSL Data Items or CSL Data, with a parenthetical statement in one place to note that this data is JSON-formatted). I propose sticking with this usage since to me it could be problematic down the road to refer to a certain type of data by the format it's encoded in. For example, pandoc-citeproc accepts CSL data formatted as YAML.

@agitter @slochower ready for re-review

slochower commented on content/02.main-text.md:
```diff
@@ -110,7 +110,9 @@ Manubot includes an additional layer of citation processing, currently unique to
 Authors can optionally define citation tags to provide short readable alternatives to the citation identifiers.
-Metadata is exported to [Citation Style Language (CSL) JSON Items](http://citeproc-js.readthedocs.io/en/latest/csl-json/markup.html#items), an open standard that is widely supported by reference managers [@doi:10.1007/978-3-319-00026-8_8; @doi:10.1080/02763869.2012.641841].
+Metadata is exported to [Citation Style Language (CSL) JSON Data Items](http://citeproc-js.readthedocs.io/en/latest/csl-json/markup.html#items), an open standard that is widely supported by reference managers [@doi:10.1007/978-3-319-00026-8_8; @doi:10.1080/02763869.2012.641841].
+However, sometimes external resources provide Manubot with invalid CSL Data, which can crash downstream citation processors, such as [pandoc-citeproc](http://hackage.haskell.org/package/pandoc-citeproc).
```
What do you think about?
```suggestion
However, sometimes external resources provide Manubot with invalid CSL Data, which can cause an error with downstream citation processors, such as [pandoc-citeproc](http://hackage.haskell.org/package/pandoc-citeproc).
```

dhimmel replied:
Does the plural not make more sense: "can cause errors"?

P.S. I am excited to try out this new apply suggestion feature.

slochower replied:
Me too! And yup, your version is better. I just wanted to get rid of "crash."

slochower commented on content/02.main-text.md:
```diff
@@ -127,10 +129,10 @@ Authors may optionally map a named tag to one of the other supported identifier
 
-Manubot formats bibliographies according to a [CSL](http://citationstyles.org/) specification.
-As a result, users can choose from thousands of existing CSL styles or use Manubot's default style.
+Manubot formats bibliographies according to a [CSL](http://citationstyles.org/) style specification.
```
I wonder if this sentence should come before the sentence "Therefore, Manubot..." which mentions CSL Data specification.

dhimmel replied:
i like the separation of CSL Data and CSL Style since the distinction can be confusing. I think we should stick with the prior paragraph referring to reference metadata and this paragraph referring to bibliographic styling.

slochower commented on content/02.main-text.md:
```diff
@@ -110,7 +110,9 @@ Manubot includes an additional layer of citation processing, currently unique to
-Metadata is exported to [Citation Style Language (CSL) JSON Items](http://citeproc-js.readthedocs.io/en/latest/csl-json/markup.html#items), an open standard that is widely supported by reference managers [@doi:10.1007/978-3-319-00026-8_8; @doi:10.1080/02763869.2012.641841].
+Metadata is exported to [Citation Style Language (CSL) JSON Data Items](http://citeproc-js.readthedocs.io/en/latest/csl-json/markup.html#items), an open standard that is widely supported by reference managers [@doi:10.1007/978-3-319-00026-8_8; @doi:10.1080/02763869.2012.641841].
+However, sometimes external resources provide Manubot with invalid CSL Data, which can crash downstream citation processors, such as [pandoc-citeproc](http://hackage.haskell.org/package/pandoc-citeproc).
+Therefore, Manubot removes invalid fields according to the [CSL Data specification](https://github.com/citation-style-language/schema).
```
```suggestion
Therefore, Manubot removes invalid fields according to the [CSL data specification](https://github.com/citation-style-language/schema).
```

agitter replied:
`CSL Data` appears three times.  If we change it here, we should change it everywhere for consistency.  Lower case `data` is okay with me if that isn't a formal proper name.

slochower committed:
Replace "crash"

dhimmel commented on content/02.main-text.md:
```diff
@@ -110,7 +110,9 @@ Manubot includes an additional layer of citation processing, currently unique to
 Authors can optionally define citation tags to provide short readable alternatives to the citation identifiers.
-Metadata is exported to [Citation Style Language (CSL) JSON Items](http://citeproc-js.readthedocs.io/en/latest/csl-json/markup.html#items), an open standard that is widely supported by reference managers [@doi:10.1007/978-3-319-00026-8_8; @doi:10.1080/02763869.2012.641841].
+Metadata is exported to [Citation Style Language (CSL) JSON Data Items](http://citeproc-js.readthedocs.io/en/latest/csl-json/markup.html#items), an open standard that is widely supported by reference managers [@doi:10.1007/978-3-319-00026-8_8; @doi:10.1080/02763869.2012.641841].
+However, sometimes external resources provide Manubot with invalid CSL Data, which can cause an error with downstream citation processors, such as [pandoc-citeproc](http://hackage.haskell.org/package/pandoc-citeproc).
```
```suggestion
However, sometimes external resources provide Manubot with invalid CSL Data, which can cause errors with downstream citation processors, such as [pandoc-citeproc](http://hackage.haskell.org/package/pandoc-citeproc).
```

dhimmel committed:
"an error" to "errors"

agitter reviewed (approved):
Ready to merge @dhimmel.

Let's confirm that the new citation `url:https://blog.dhimmel.com/citation-styles/` worked correctly after this version builds.
